### PR TITLE
draft implementation of tables pregeneration

### PR DESCRIPTION
**Pull request by pavel-collab**
This is a my implementation of tables pregeneration. I've just move all generation functions to the new file generation/GenerateTables.c. This file is compiled, run and write all information to the 3 .dat files. After that content from this files is copied and pasted to the config.h.

Also, For now I've commented some functions in the original files, cause now we don't need it anymore.

As I've understood we need to generate this tables once at all. We will never generate it again (except if we won't change the base polynom).

### CODE REVIEW

File: generation/GenerateTables.c [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  for (int i = 0; i < BLOCK_SIZE; ++i)
+    for (int j = 0; j < 256; ++j)
+      for (int k = 0; k < BLOCK_SIZE; ++k)
+        LS_tbl[i][j][k] ^= poly_mul(S[j], l_tbl[k][i]);
+}
+
+int main(int argc, char* argv[]) {
+    const char* coefs_file_path = "./coefs.dat";
```
Comment (Exactlywb) [inline generation/GenerateTables.c:107]: Do not hardcode the pathes to output files.  Take it from `argv` parameter.

File: generation/GenerateTables.c [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+              fprintf(fd1, "0x%x}", Muls[i][j]);
+            } else {
+              fprintf(fd1, "0x%x, ", Muls[i][j]);
+            }
+          }
+          fprintf(fd1, "\n");
+      }
+      fclose(fd1);
+    } else {
+      perror("Error open file\n");
```
Comment (Exactlywb) [inline generation/GenerateTables.c:134]: Remove '\n' from perror. Perror output looks like this:
```
Error open file: error_message_here
```

But now you have '\n' and getting:
```
Error open file:
error_message_here
```

File: generation/GenerateTables.c [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    }
+
+    FILE* fd2 = fopen(ls_tbl_file_path, "a");
```
Comment (Exactlywb) [inline generation/GenerateTables.c:137]: Why mode is 'a'? ISTM it would be 'w'.

File: lib/grasshopper/grasshopper.c [hunk 4]
```diff
@@ -153,18 +153,33 @@ void OFB_decryption(Block *text, Key *key, const Block *iv, size_t NBlocks) {
   }
 }
 
-// TODO: autogenerate this and move to config.h
+// // TODO: autogenerate this and move to config.h
 Block Coefs[NUM_ROUNDS / 2 - 1][BIT_IN_BYTES];
```
Comment (Exactlywb) [inline lib/grasshopper/grasshopper.c:157]: Pregenerate 'Coefs' too.

File: lib/grasshopper/grasshopper.c [hunk 5]
```diff
@@ -153,18 +153,33 @@ void OFB_decryption(Block *text, Key *key, const Block *iv, size_t NBlocks) {
   }
 }
 
-// TODO: autogenerate this and move to config.h
+// // TODO: autogenerate this and move to config.h
```
Comment (Exactlywb) [inline lib/grasshopper/grasshopper.c:156]: Remove TODO. It's necessary no more.

File: lib/grasshopper/grasshopper.c [hunk 6]
```diff
@@ -153,18 +153,33 @@ void OFB_decryption(Block *text, Key *key, const Block *iv, size_t NBlocks) {
...
-unsigned char LS_tbl[16][256][16];
+// unsigned char Muls[256][16];
+// unsigned char LS_tbl[16][256][16];
+
+
+/**
+ * By generator we can only generate plain data, but not BLock structure.
+ * This function produce an array of structures Block from given data.
+*/
+void make_coefs_from_data(const unsigned char coefs_data[][BIT_IN_BYTES][BLOCK_SIZE]) {
```
Comment (Exactlywb) [inline lib/grasshopper/grasshopper.c:166]: As I've already written: pregenerate 'Coefs' too and remove this function.

File: lib/grasshopper/grasshopper.c [hunk 7]
```diff
@@ -273,52 +288,52 @@ static unsigned char poly_mul(unsigned char left, unsigned char right) {
...
-    memcpy(l_tbl, tmp, 256 * sizeof(unsigned char));
-    memset(tmp, 0, 256 * sizeof(unsigned char));
-  }
-
-  for (int i = 0; i < BLOCK_SIZE; ++i)
-    for (int j = 0; j < 256; ++j)
-      for (int k = 0; k < BLOCK_SIZE; ++k)
-        LS_tbl[i][j][k] ^= poly_mul(S[j], l_tbl[k][i]);
-}
+// /* Pregenerations.  */
```
Comment (Exactlywb) [inline lib/grasshopper/grasshopper.c:291]: Remove this functions (which are comments now).

Moreover, in the future, if you'd like to switch the code off it's better to use preprocessor directives. e.g.:

```
/*
DEAD CODE COULD BE HERE BUT THIS IS NOT THE BEST WAY
*/
#if 0
DEAD CODE COULD BE HERE AND THIS IS THE BEST WAY
#endif
```

File: tools/simple/simple.c [hunk 8]
```diff
@@ -10,19 +10,19 @@ void apply_x(Block *block, const Block *key) {
     block->data[i] = block->data[i] ^ key->data[i];
 }
 
-void apply_l(Block *block) {
-  for (int i = 0; i < BLOCK_SIZE; ++i) {
-    unsigned char tmp = 0;
-    for (int j = 0; j < BLOCK_SIZE; ++j)
-      tmp ^= Muls[block->data[j]][j];
+// void apply_l(Block *block) {
```
Comment (Exactlywb) [inline tools/simple/simple.c:13]: Ditto

File: tools/vectorized/vectorized.c [hunk 9]
```diff
@@ -20,6 +20,13 @@ void apply_x(Block *block, const Block *key) {
 }
 
 // TODO: try to get a profit here
+/**
+ * From pavel.
```
Comment (Exactlywb) [inline tools/vectorized/vectorized.c:24]: pavel -> 'Pavel'

**Review by Exactlywb — COMMENTED** [review]
Looks good to me. Pls test that everything works fine (now I'm not able to do this). And fix the next issues:

**Comment by pavel-collab** [comment]
I've generated a tables.h and put it into include/utils/. For now I need your help.
@Exactlywb could you, please check out GenerateTables.c and other files once again (code style fixes etc).
@kolobabka I need your help with includes and build system. As you were writing CMake files, could you, please, make tables.h available in  grasshopper.c, config.h and grasshopper.h and maybe other files if it need and write necessary dependencies.

And please, don't worry about commits. When all fixes will be done, I will do patch to save all changes in one commit.

**Comment by Exactlywb** [comment]
LGTM. Needs to be tested (also perfomance tests)

File: include/utils/config.h [hunk 10]
```diff
@@ -14,40 +14,12 @@
...
-    0x30, 0x37, 0x6B, 0xE4, 0x88, 0xD9, 0xE7, 0x89, // C8..CF
-    0xE1, 0x1B, 0x83, 0x49, 0x4C, 0x3F, 0xF8, 0xFE, // D0..D7
-    0x8D, 0x53, 0xAA, 0x90, 0xCA, 0xD8, 0x85, 0x61, // D8..DF
-    0x20, 0x71, 0x67, 0xA4, 0x2D, 0x2B, 0x09, 0x5B, // E0..E7
-    0xCB, 0x9B, 0x25, 0xD0, 0xBE, 0xE5, 0x6C, 0x52, // E8..EF
-    0x59, 0xA6, 0x74, 0xD2, 0xE6, 0xF4, 0xB4, 0xC0, // F0..F7
-    0xD1, 0x66, 0xAF, 0xC2, 0x39, 0x4B, 0x63, 0xB6  // F8..FF
-};
+/*
+DEAD CODE
```
Comment (Exactlywb) [inline include/utils/config.h:18]: Remove this spurious comment from here.

File: generation/GenerateTables.c [hunk 11]
```diff
@@ -1,9 +1,58 @@
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
-
-#include <utils/config.h>
-#include <grasshopper/grasshopper.h>
+#include <stddef.h>
+
+// #include <utils/config.h>
```
Comment (Exactlywb) [inline generation/GenerateTables.c]: Do you have a plan how to include these headers or this should be deleted?

File: generation/GenerateTables.c [hunk 12]
```diff
@@ -26,7 +75,7 @@ void block_dump_to_file(const Block *block, FILE* fd) {
   }
 }
 
-void apply_l(Block *block, unsigned char Muls[][16]) {
+void  apply_l(Block *block, unsigned char Muls[][16]) {
```
Comment (Exactlywb) [inline generation/GenerateTables.c:75]: Trailing whitespaces.

File: generation/GenerateTables.c [hunk 13]
```diff
@@ -113,15 +162,18 @@ void write_muls_to_file(FILE* fd, unsigned char Muls[][16]) {
 
   fprintf(fd, "static unsigned Muls[256][16] = {\n");
   for (size_t i = 0; i < 256; ++i) {
+    fprintf(fd, "{\n");
       for (size_t j = 0; j < 16; ++j) {
         if (i == 255 && j == 15) {
-          fprintf(fd, "0x%x};", Muls[i][j]);
+          fprintf(fd, "0x%x}\n};", Muls[i][j]);
```
Comment (Exactlywb) [inline generation/GenerateTables.c:165]: Use `%#x` in printf functions.

File: generation/GenerateTables.c [hunk 14]
```diff
@@ -131,19 +183,24 @@ void write_ld_tbl_to_file(FILE* fd, unsigned char LS_tbl[][256][16]) {
...
   fprintf(fd, "static unsigned char LS_tbl[16][256][16] = {\n");
-  for (size_t i = 0; i < 16; ++i) {
-      for (size_t j = 0; j < 256; ++j) {
-        for (size_t k = 0; k < 16; ++k) {
+  for (size_t i = 0; i < 16; i++) {
+    fprintf(fd, "{\n");
+      for (size_t j = 0; j < 256; j++) {
+        fprintf(fd, "{");
+        for (size_t k = 0; k < 16; k++) {
           if (i == 15 && j == 255 && k ==15) {
```
Comment (Exactlywb) [inline generation/GenerateTables.c:188]: Emit whitespace in `k == 15`

File: generation/GenerateTables.c [hunk 15]
```diff
@@ -131,19 +183,24 @@ void write_ld_tbl_to_file(FILE* fd, unsigned char LS_tbl[][256][16]) {
...
-            fprintf(fd, "0x%x};", LS_tbl[i][j][k]);
+            fprintf(fd, "0x%x}\n}\n};", LS_tbl[i][j][k]);
+          } else if (j == 255 && k ==15) {
+            fprintf(fd, "0x%x}\n},\n", LS_tbl[i][j][k]);
+          } else if (k ==15) {
```
Comment (Exactlywb) [inline generation/GenerateTables.c:192]: Ditto.

File: generation/GenerateTables.c [hunk 16]
```diff
@@ -131,19 +183,24 @@ void write_ld_tbl_to_file(FILE* fd, unsigned char LS_tbl[][256][16]) {
...
-        for (size_t k = 0; k < 16; ++k) {
+  for (size_t i = 0; i < 16; i++) {
+    fprintf(fd, "{\n");
+      for (size_t j = 0; j < 256; j++) {
+        fprintf(fd, "{");
+        for (size_t k = 0; k < 16; k++) {
           if (i == 15 && j == 255 && k ==15) {
-            fprintf(fd, "0x%x};", LS_tbl[i][j][k]);
+            fprintf(fd, "0x%x}\n}\n};", LS_tbl[i][j][k]);
+          } else if (j == 255 && k ==15) {
```
Comment (Exactlywb) [inline generation/GenerateTables.c:190]: Ditto.

File: tools/vectorized/vectorized.c [hunk 17]
```diff
@@ -1,10 +1,6 @@
-#include "grasshopper.h"
 #include <string.h>
-
+#include "grasshopper.h"
```
Comment (Exactlywb) [inline tools/vectorized/vectorized.c:2]: Include 'grasshopper.h' before stdlibs.

**Comment by Exactlywb** [comment]
Also update README.md with table generation info pls

File: tools/vectorized/vectorized.c
Comment (pavel-collab) [inline tools/vectorized/vectorized.c:1]: You still have function apply_l in this file. It's not need here anymore. This function is used only for tables content generation and you have moved it to GenerateTables.c
So, you need to remove its implementation from this file. Also check if it's still exists this function prototype in some .h files and remove it also.

**Review by pavel-collab — COMMENTED** [review]
You still have function apply_l in vectorized.c. It's not need here anymore. This function is used only for tables content generation and you have moved it to GenerateTables.c
So, you need to remove its implementation from this file. Also check if it's still exists this function prototype in some .h files and remove it also.

**Comment by Exactlywb** [comment]
> You still have function apply_l in vectorized.c. It's not need here anymore. This function is used only for tables content generation and you have moved it to GenerateTables.c So, you need to remove its implementation from this file. Also check if it's still exists this function prototype in some .h files and remove it also.

Done (performance boost for this function is still available in README.md; it is too sad to remove it)

**Comment by pavel-collab** [comment]
> > You still have function apply_l in vectorized.c. It's not need here anymore. This function is used only for tables content generation and you have moved it to GenerateTables.c So, you need to remove its implementation from this file. Also check if it's still exists this function prototype in some .h files and remove it also.
> 
> Done (performance boost for this function is still available in README.md; it is too sad to remove it)

It will be removed in the branch with experiment python script. When branch experiment will be merged in master user will be able to generate such table automaticly. It's not a problem for this branch now.

File: generation/GenerateTables.c [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+  for (int i = 0; i < BLOCK_SIZE; ++i) {
+    if (i == BLOCK_SIZE-1) {
+      fprintf(fd, "%#x} ", block->data[i]);
+    } else {
+      fprintf(fd, "%#x, ", block->data[i]);
+    }
+  }
+}
+
+void apply_l(Block *block, unsigned char Muls[][16]) {
```
Comment (kolobabka) [inline generation/GenerateTables.c:76]: static maybe?

**Pull request merged**